src/channel.cr

```diff
@@ -102,82 +114,126 @@ class Channel(T)
     Closed
   end
 
-  private record Sender(T), fiber : Fiber, value : T, state_ptr : DeliveryState*, select_context : SelectContext(Nil)?
-  private record Receiver(T), fiber : Fiber, value_ptr : T*, state_ptr : DeliveryState*, select_context : SelectContext(T)?
+  # :nodoc:
+  struct Sender(T)
+    include Crystal::PointerLinkedList::Node
+    property fiber : Fiber
+    property data : T
+    property state : DeliveryState
+    property select_context : SelectContext(Nil)?
+
+    def initialize
+      @fiber = uninitialized Fiber
+      @data = uninitialized T
+      @state = DeliveryState::None
+    end
+  end
+
+  # :nodoc:
+  struct Receiver(T)
+    include Crystal::PointerLinkedList::Node
+    property fiber : Fiber
+    property data : T
+    property state : DeliveryState
+    property select_context : SelectContext(T)?
+
+    def initialize
+      @fiber = uninitialized Fiber
+      @data = uninitialized T
+      @state = DeliveryState::None
+    end
+  end
 
   def initialize(@capacity = 0)
     @closed = false
-    @senders = Deque(Sender(T)).new
-    @receivers = Deque(Receiver(T)).new
+
+    @senders = Crystal::PointerLinkedList(Sender(T)).new
+    @receivers = Crystal::PointerLinkedList(Receiver(T)).new
+
     if capacity > 0
       @queue = Deque(T).new(capacity)
     end
   end
 
-  def close
+  private macro close_sender_receiver(ptr)
```

Comment:
Why a `macro` instead of regular `def`?

Reply:
I think there is no difference between `macro` and `def` in this case. Just prefer `macro`.

Reply:
Macros should be used only when necessary. I'd suggest using `def` instead.

Reply:
OK, I see. I will change that. BTW, I have a question about macro use. For example, [this](https://github.com/crystal-lang/crystal/blob/5a5704c4dd446be7570f8a39e9e552578c462ac8/src/debug/dwarf/line_numbers.cr#L264). Is it necessary or just a bad example?

Reply:
@ysbaddaden will know the best, since he wrote this piece of code.

Reply:
Macros should be avoided unless you need to generate code, yet macros can be a nice alternative to defs in some occasions.

In the DWARF::LineNumbers example, `registers` is a struct so we can avoid some HEAP allocations, but we can't pass the struct as a method argument because it would pass it by value (copy it) when we want `registers` to be modified in-place. I could pass a pointer, but pointers are meh in Crystal (and rightfully so). We could pass the struct and return it back, but it's boring, there is no guarantee that LLVM will optimize the copies away (or even inline the method).

I.e. the macro avoids a repetition, HEAP allocations and/or a pointer. I believe it's acceptable.

Reply:
@ysbaddaden I understand now. Thank you.